core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinNoKeyFastRecordCursorFactory.java

```diff
@@ -0,0 +1,369 @@
+                int columnSplit,
+                Record nullRecord,
+                int masterTimestampIndex,
+                int slaveTimestampIndex,
+                int lookahead
+        ) {
+            this.columnSplit = columnSplit;
+            this.record = new OuterJoinRecord(columnSplit, nullRecord);
+            this.masterTimestampIndex = masterTimestampIndex;
+            this.slaveTimestampIndex = slaveTimestampIndex;
+            this.lookahead = lookahead;
+        }
+
+        @Override
+        public void calculateSize(SqlExecutionCircuitBreaker circuitBreaker, Counter counter) {
+            masterCursor.calculateSize(circuitBreaker, counter);
+        }
+
+        @Override
+        public void close() {
+            masterCursor = Misc.free(masterCursor);
+            slaveCursor = Misc.free(slaveCursor);
+        }
+
+        @Override
+        public Record getRecord() {
+            return record;
+        }
+
+        @Override
+        public SymbolTable getSymbolTable(int columnIndex) {
+            if (columnIndex < columnSplit) {
+                return masterCursor.getSymbolTable(columnIndex);
+            }
+            return slaveCursor.getSymbolTable(columnIndex - columnSplit);
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (isMasterHasNextPending) {
+                masterHasNext = masterCursor.hasNext();
+                isMasterHasNextPending = false;
+            }
+            if (masterHasNext) {
+                final long masterTimestamp = masterRecord.getTimestamp(masterTimestampIndex);
+                if (masterTimestamp < slaveTimestamp) {
+                    isMasterHasNextPending = true;
+                    return true;
+                }
+                nextSlave(masterTimestamp);
+                isMasterHasNextPending = true;
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public SymbolTable newSymbolTable(int columnIndex) {
+            if (columnIndex < columnSplit) {
+                return masterCursor.newSymbolTable(columnIndex);
+            }
+            return slaveCursor.newSymbolTable(columnIndex - columnSplit);
+        }
+
+        @Override
+        public long size() {
+            return masterCursor.size();
+        }
+
+        @Override
+        public void toTop() {
+            slaveTimestamp = Long.MIN_VALUE;
+            slaveFrameIndex = -1;
+            slaveFrameRow = Long.MIN_VALUE;
+            record.hasSlave(false);
+            masterCursor.toTop();
+            slaveCursor.toTop();
+            isMasterHasNextPending = true;
+            isSlaveOpenPending = false;
+            isSlaveForwardScan = true;
+        }
+
+        private boolean linearScan(TimeFrame frame, long masterTimestamp) {
+            final long scanHi = Math.min(slaveFrameRow + lookahead, frame.getRowHi());
+            while (slaveFrameRow < scanHi) {
```

Comment:
You may want to stay on linear search if there are a lot of matching timestamps, binary search will not be any faster
```suggestion
            while (slaveFrameRow < scanHi || (slaveTimestamp == masterTimestamp && slaveFrameRow < frame.getRowHi())) {
```

Reply:
Good idea. Applied in b89b1e7645868bddefd4d13d3ebce4f6a9d8ea28